### PR TITLE
federation: Adding federation controller manager to federation e2es

### PR DESCRIPTION
**Pull request by nikhiljindal**
Adding federation controller manager to federation e2es.
Havent been able to verify that this works: https://github.com/kubernetes/kubernetes/pull/25125#issuecomment-222860475

Followed the same patter as federation-apiserver and made the changes at same places, so hopefully should work :)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
cc @kubernetes/sig-cluster-federation @colhom @mfanjie

File: federation/manifests/federation-controller-manager-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        command:
+        - /usr/local/bin/federation-controller-manager
+        - --master={{.FEDERATION_API_HOST}}:{{.FEDERATION_API_PORT}}
+        ports:
+        - containerPort: 443
+          name: https
+        - containerPort: 8080
+          name: local
+        env:
+        - name: POD_NAMESPACE
```
Comment (nikhiljindal) [inline federation/manifests/federation-controller-manager-deployment.yaml]: Adding POD_NAMESPACE env var as per https://github.com/kubernetes/kubernetes/pull/25125#issuecomment-221444046

File: federation/manifests/federation-apiserver-deployment.yaml [hunk 2]
```diff
@@ -23,7 +23,6 @@ spec:
         - --service-cluster-ip-range={{.FEDERATION_SERVICE_CIDR}}
         - --secure-port=443
         - --advertise-address={{.FEDERATION_API_HOST}}
-        - --admission-control=NamespaceLifecycle,LimitRanger,SecurityContextDeny,ServiceAccount,ResourceQuota
```
Comment (nikhiljindal) [inline federation/manifests/federation-apiserver-deployment.yaml]: We are not supporting these admission controls in 1.3. Ref https://github.com/kubernetes/kubernetes/pull/26298#issuecomment-221697124

**Comment by quinton-hoole** [comment]
cc: @madhusudancs

File: federation/manifests/federation-controller-manager-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      labels:
+        app: federated-cluster
+        module: federation-controller-manager
+    spec:
+      containers:
+      - name: controller-manager
+        image: {{.FEDERATION_CONTROLLER_MANAGER_IMAGE_REPO}}:{{.FEDERATION_CONTROLLER_MANAGER_IMAGE_TAG}}
+        command:
+        - /usr/local/bin/federation-controller-manager
+        - --master={{.FEDERATION_API_HOST}}:{{.FEDERATION_API_PORT}}
```
Comment (colhom) [inline federation/manifests/federation-controller-manager-deployment.yaml]: This is the publicly exposed endpoint, either via nodeport or lb depending on provider. Instead you should refer here to the federated-apiserver service by it's cluster DNS name: 

`--master=https://{{.FEDERATION_APISERVER_DEPLOYMENT_NAME}}:443`

File: federation/cluster/common.sh [hunk 4]
```diff
@@ -123,6 +129,24 @@ function create-federated-api-objects {
 	sleep 4
     done
 
+    # Verify that federation-controller-manager pod is running.
+    for i in {1..30};do
+	#TODO(colhom): in the future this needs to scale out for N pods. This assumes just one pod
```
Comment (colhom) [inline federation/cluster/common.sh]: We also need a generic way of expressing "wait until these pods are all running/ready", along with timeout/max retry logic

Comment (colhom) [inline federation/cluster/common.sh]: ref #1899

File: federation/cluster/common.sh [hunk 5]
```diff
@@ -137,7 +161,7 @@ function create-federated-api-objects {
 # FEDERATION_IMAGE_TAG: push all federated images with this tag. Used for ci testing
 function push-federated-images {
     : "${FEDERATION_PUSH_REPO_BASE?Must set FEDERATION_PUSH_REPO_BASE env var}"
-    local FEDERATION_BINARIES=${FEDERATION_BINARIES:-'federation-apiserver'}
+    local FEDERATION_BINARIES=${FEDERATION_BINARIES:-'federation-apiserver,federation-controller-manager'}
```
Comment (colhom) [inline federation/cluster/common.sh]: `FEDERATION_BINARIES` should be a space-separated list, not comma.

Comment (madhusudancs) [inline federation/cluster/common.sh]: +1

Some of the fixes to this PR are here - https://github.com/kubernetes/kubernetes/pull/26636/commits/2bd99740786fef727bb177df69a4d53ddf47cba2 Feel free to cherrypick the commit or just copy the changes.

File: federation/manifests/federation-controller-manager-deployment.yaml (hunk 3 above)
Comment (nikhiljindal) [inline federation/manifests/federation-controller-manager-deployment.yaml]: fixed. Also added a YAML to create a ClusterIP service.

File: federation/cluster/common.sh (hunk 5 above)
Comment (nikhiljindal) [inline federation/cluster/common.sh]: thanks, fixed

File: federation/cluster/common.sh (hunk 4 above)
Comment (nikhiljindal) [inline federation/cluster/common.sh]: ack

**Comment by nikhiljindal** [comment]
Fixed code as per comments and pushed a new version, PTAL

File: federation/cluster/common.sh [hunk 6]
```diff
@@ -47,6 +47,10 @@ function create-federated-api-objects {
     export FEDERATION_APISERVER_IMAGE_REPO="${FEDERATION_PUSH_REPO_BASE}/federation-apiserver"
     export FEDERATION_APISERVER_IMAGE_TAG="${FEDERATION_IMAGE_TAG:-$(cat ${KUBE_ROOT}/_output/${KUBE_BUILD_STAGE}/server/${KUBE_PLATFORM}-${KUBE_ARCH}/kubernetes/server/bin/federation-apiserver.docker_tag)}"
 
+    export FEDERATION_CONTROLLER_MANAGER_DEPLOYMENT_NAME="federation-deployment-manager"
```
Comment (madhusudancs) [inline federation/cluster/common.sh]: "federation-controller-manager" you mean?

Comment (nikhiljindal) [inline federation/cluster/common.sh]: yes, fixed

**Comment by nikhiljindal** [comment]
Updated code as per comments.
Also squashed commits. PTAL.

**Comment by madhusudancs** [comment]
LGTM.

**Pull request merged**